### PR TITLE
python310Packages.pyperscan: init at 0.2.2

### PR DESCRIPTION
**Pull request by tnias**
Required for https://github.com/NixOS/nixpkgs/issues/217836

###### Description of changes

[pyperscan](https://github.com/vlaci/pyperscan) is a [hyperscan](https://github.com/intel/hyperscan) binding for Python, which supports [vectorscan](https://github.com/VectorCamp/vectorscan). This PR also adds the mentioned vectorscan library.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tnias** [comment]
I currently have a very restrictive platforms list (x86_64 and aarch64 linux).

@vlaci Do you have an idea if darwin on x86_64 and aarch64 work? Also, do you want to be added to the maintainers list of those packages?

**Comment by vlaci** [comment]
> 

It should work on x86_64 and aarch64 as well, both on Linux and Darwin. I'd be happy to be maintainers of these packages.
I also have a derivation somewhere for vectorscan. I had a hard time getting it linked properly. If pyperscan's test passes, than it works.

There is an SVE (akin to AVX on ARM) related build option we could enable. I'll check it later .

**Comment by tnias** [comment]
Looks like `BUILD_SVE` (and `BUILD_SVE2`) are defined in [cmake/arch.cmake](https://github.com/VectorCamp/vectorscan/blob/vectorscan/5.4.9/cmake/arch.cmake#L40-L62). The `HAVE_SVE`  flag needs to be set.

[CMakeLists.txt](https://github.com/VectorCamp/vectorscan/blob/vectorscan/5.4.9/CMakeLists.txt) tries to be smart and guess a good `GNUCC_ARCH`, this might break because of our wrapped compiler shenanigans.
I think for `HAVE_SVE` detection we have to set the effective march/mtune gcc flags.

File: pkgs/development/libraries/vectorscan/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+      Vectorscan will follow Intel’s API and internal algorithms where possible,
+      but will not hesitate to make code changes where it is thought of giving
+      better performance or better portability. In addition, the code will be
+      gradually simplified and made more uniform and all architecture specific
+      code will be abstracted away.
+    '';
+    homepage = "https://www.vectorcamp.gr/vectorscan/";
+    changelog = "https://github.com/VectorCamp/vectorscan/blob/${src.rev}/CHANGELOG.md";
+    platforms = [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vectorscan/default.nix:58]:
```suggestion
    platforms = platforms.unix;
```
more or less for our purposes

File: pkgs/development/libraries/vectorscan/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ragel
+    util-linux
+    python3
+  ];
+
+  buildInputs = [
+    boost
+  ];
+
+  cmakeFlags = lib.optional (enableShared) [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vectorscan/default.nix:34]:
```suggestion
  cmakeFlags = lib.optionals enableShared [
```

File: pkgs/development/libraries/vectorscan/default.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    "-DBUILD_STATIC_AND_SHARED=ON"
+  ] ++ lib.optional (stdenv.hostPlatform.system == "x86_64-linux") [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vectorscan/default.nix:36]:
```suggestion
  ] ++ lib.optionals (stdenv.hostPlatform.system == "x86_64-linux") [
```

File: pkgs/development/libraries/vectorscan/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    "-DBUILD_AVX2=ON"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vectorscan/default.nix:37]: Use stdenv.hostPlatform.avx2Support for that

File: pkgs/development/libraries/vectorscan/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    "-DBUILD_AVX512=ON"
+    "-DFAT_RUNTIME=ON"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vectorscan/default.nix:39]: and stdenv.hostPlatform.avx512Support

File: pkgs/development/libraries/vectorscan/default.nix (hunk 1 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix:58]: I tested with i686-linux and it does not work. I assume, that many platforms in platforms.unix are not suppored by vectorscan. Should I still use platforms.unix?

File: pkgs/development/libraries/vectorscan/default.nix (hunk 5 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix]: done

File: pkgs/development/libraries/vectorscan/default.nix (hunk 4 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix]: done

File: pkgs/development/libraries/vectorscan/default.nix (hunk 3 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix]: after reorganizing i opted to use lib.optional for single options

File: pkgs/development/libraries/vectorscan/default.nix (hunk 2 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix]: after reorganizing i opted to use lib.optional for single options

**Comment by tnias** [comment]
Result of `nixpkgs-review pr 232314` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>python310Packages.pyperscan</li>
    <li>python310Packages.pyperscan.dist</li>
    <li>python311Packages.pyperscan</li>
    <li>python311Packages.pyperscan.dist</li>
    <li>vectorscan</li>
  </ul>
</details>

Result of `nixpkgs-review pr 232314` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>python310Packages.pyperscan</li>
    <li>python310Packages.pyperscan.dist</li>
    <li>python311Packages.pyperscan</li>
    <li>python311Packages.pyperscan.dist</li>
    <li>vectorscan</li>
  </ul>
</details>

File: pkgs/development/libraries/vectorscan/default.nix (hunk 1 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix]: The darwin builds in ofborg fail. => Only x86_64 and aarch64 left.

Comment (SuperSandro2000) [reply pkgs/development/libraries/vectorscan/default.nix]: This is more or less about what hydra should build. Most programs that have unix probably don't support everything in there and that is alright.

If darwin does not work, we should set this to platforms.linux. If also aarch64 does not work, we should set it to ``[ "x86_64-linux" ]``

Comment (vlaci) [reply pkgs/development/libraries/vectorscan/default.nix]: Darwin should work, given that vectorscan is available

Comment (vlaci) [reply pkgs/development/libraries/vectorscan/default.nix]: Darwin should work, as I have binary wheels targeting that platform. I am unsure what the issue is

Comment (vlaci) [reply pkgs/development/libraries/vectorscan/default.nix]: From the [logs](https://logs.ofborg.org/?attempt_id=6f3aebce-da7b-4f31-b526-899927644f1a&key=nixos%2Fnixpkgs.232314) I think fat runtime defaults to enabled, this causes the build errors IMHO
I explicitly disable it on darwin, when build statically: https://github.com/vlaci/pyperscan/blob/e8073597ef9532878d9209acf68c97f5a7433985/hyperscan-sys/build.rs#L84

**Comment by vlaci** [comment]
Based on the [logs](https://logs.ofborg.org/?attempt_id=6f3aebce-da7b-4f31-b526-899927644f1a&key=nixos%2Fnixpkgs.232314), the issue originates from [build_wrapper.sh](https://github.com/VectorCamp/vectorscan/blob/b4bba94b1a250603b0b198e0394946e32f6c3f30/cmake/build_wrapper.sh), which contains a bunch of GNU toolchain dependent stuff, and dependencies on other GNU tools. 
I think fat runtime defaults to enabled, as this script is used only on [that codepath](https://github.com/VectorCamp/vectorscan/blob/b4bba94b1a250603b0b198e0394946e32f6c3f30/CMakeLists.txt#L1277).
This causes the build errors IMHO. 

I explicitly disable it on darwin, when build statically: https://github.com/vlaci/pyperscan/blob/e8073597ef9532878d9209acf68c97f5a7433985/hyperscan-sys/build.rs#L84

File: pkgs/development/libraries/vectorscan/default.nix [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  nativeBuildInputs = [
+    cmake
+    ragel
+    util-linux
+    python3
+  ];
+
+  buildInputs = [
+    boost
+  ];
```
[the remark this answers is not in the document]
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix:32]: The error occurs when linking stuff in pyperscan, so libiconv should be added to pyperscan, no?

[the remark this answers is not in the document]
Comment (vlaci) [reply pkgs/development/libraries/vectorscan/default.nix:32]: You are right, my bad

[the remark this answers is not in the document]
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix:32]: After adding libiconv ofborg is happy. => resolving this thread

File: pkgs/development/libraries/vectorscan/default.nix (hunk 1 above)
Comment (tnias) [reply pkgs/development/libraries/vectorscan/default.nix]: Darwin now works and we have platforms.unix. => resolving this thread

File: pkgs/development/python-modules/pyperscan/default.nix [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  src = fetchFromGitHub {
+    owner = "vlaci";
+    repo = "pyperscan";
+    rev = "v${version}";
+    hash = "sha256-ioNGEmWy+lEzazF1RzMFS06jYLNYll3QSlWAF0AoU7Y=";
+  };
+
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+  };
```
Comment (vlaci) [inline pkgs/development/python-modules/pyperscan/default.nix:25]: I am not sure, what is the best practice in nixpkgs, I see examples 32 for `importCargoLock` but 81 for `fetchCargoTarball`

**Review by vlaci — COMMENTED** [review]
I think these are great changes. I just have one question inline.

**Review by vlaci — APPROVED** [review]

**Pull request merged**